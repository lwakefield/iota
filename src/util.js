export function observe (obj, fn) {
  if (!obj) return obj
  if (obj.__observer__) return obj.__observer__

  const proxy = new Proxy(obj, {
    set (target, property, val) {
      target[property] = val
      fn()
      return true
    },
    get (target, property) {
      const result = target[property]
      return typeof result === 'object' && property !== '__observer__' ?
        observe(result, fn) :
        result
    }
  })

  Object.defineProperty(proxy, '__observer__', {value: proxy})
  return proxy
}

export function proxy (ontoObj, val) {
  if (!val) return

  for (let key of Object.keys(val)) {
    Object.defineProperty(ontoObj, key, {
      enumerable: true,
      configurable: true,
      get () {
        return val[key]
      },
      set (newVal) {
        val[key] = newVal
      },
    })
  }
}

export function arrToObj (arr, fn) {
  return arr.reduce(
    (result, v) => Object.assign(result, fn(v)),
    {}
  )
}

export function objMap (obj, fn) {
  return Object.keys(obj).reduce(
    (result, k) => Object.assign(result, {[k]: fn(obj[k], k)}),
    {}
  )
}

export const swap = (arr, a, b) => {
  if (!a || !b) return
  const [aIndex, bIndex] = [arr.indexOf(a), arr.indexOf(b)]
  arr.splice(aIndex, 1, b)
  arr.splice(bIndex, 1, a)
}

export const max = (a,b) => a > b ? a : b
