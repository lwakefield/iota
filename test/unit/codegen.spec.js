/* eslint-env jest */
import beautify from 'js-beautify'

import {
  codegen,
  codegenOptions,
  codegenTextNode,
  codegenChildren,
  codegenElementNode,
} from '../../src/codegen'
import Directive, {
  registerDirective,
  unregisterDirective,
} from '../../src/directives'
import {htoe} from './util'

const assertCode = code => expect(beautify.js(code)).toMatchSnapshot()

describe('Codegen', () => {
  describe('codegen', () => {
    it('generates a simple app', () => {
      const root = htoe(`
        <div id="app">
          <h1> Hello \${username} </h1>
          <ul>
            <li i-for="m of msgs" i-if="m.show">
              \${m.text}
            </li>
          </ul>
        </div>
      `)
      const app = codegen(root)
      expect(app).toBeInstanceOf(Function)

      // const rendered = app.call({
      //   vnode,
      //   tnode,
      //   attr,
      //   event,
      //   username: 'foobar',
      //   msgs: [
      //     {text: 'one', show: true},
      //     {text: 'two', show: false},
      //     {text: 'three', show: true},
      //   ]
      // })
      // TODO: fix this
      // expect(rendered).to.eql(
      //   vnode('div', {attributes: {id: 'app'}}, [
      //     vnode('h1', {}, [tnode(' Hello foobar ')]),
      //     vnode('ul', {}, [
      //       vnode('li', {key: 0}, [tnode('one')]),
      //       null,
      //       vnode('li', {key: 2}, [tnode('three')])
      //     ])
      //   ])
      // )
    })
  })

  describe('codegenOptions', () => {
    const cg = html => codegenOptions(htoe(html))

    it('generates multiple attributes', () => {
      assertCode(cg('<p id="foo" class="class1 class2 ${class3}"></p>'))
    })
    it('generates props', () => {
      assertCode(cg('<p :foo="foobar"></p>'))
    })
    it('generates events', () => {
      assertCode(cg('<p @input="handleInput"></p>'))
      assertCode(cg('<p @input="handleInput(id)"></p>'))
      assertCode(cg('<input @input="handleInput" value="${name}"></input>'))
    })
    it('generates directives', () => {
      class Foo extends Directive {}
      registerDirective(Foo)

      assertCode(cg('<p foo></p>'))
      assertCode(cg('<p foo="bar"></p>'))

      unregisterDirective(Foo)
    })
  })

  describe('codegenTextNode', () => {
    it('generates tnode', () => {
      const t = htoe('foo ${bar}')
      assertCode(codegenTextNode(t))
    })
  })

  describe('codegenChildren', () => {
    it('generates children', () => {
      const el = htoe('<div><p></p><span></span><h1></h1></div>')
      assertCode(codegenChildren(el))
    })
  })

  describe('codegenElement', () => {
    it('generates for an element with no children and no attributes', () => {
      assertCode(codegenElementNode(htoe('<div />')))
    })
    it('generates for an element with attributes and no children', () => {
      const el = htoe(
        '<div id="foo" class="class1 class2 ${class3}"/>'
      )
      const code = codegenElementNode(el)
      assertCode(code)
    })
    it('generates for an element with children and no attributes', () => {
      const el = htoe('<div><p></p><span></span><h1></h1></div>')
      assertCode(codegenElementNode(el))
    })
    it('generates for an element with i-if', () => {
      const el = htoe('<div i-if="toggle" />')
      assertCode(codegenElementNode(el))
    })
    it('generates for an element with i-for', () => {
      const el = htoe('<div i-for="val of vals" />')
      assertCode(codegenElementNode(el))
    })
    it('generates for an element with i-for and i-if', () => {
      const el = htoe('<div i-for="val of vals" i-if="toggle" />')
      assertCode(codegenElementNode(el))
    })
    it('generates for an element with children and attributes', () => {
      const el = htoe(
        `<div id="foo" class="class1 class2 \${class3}">
          <p></p><span></span><h1></h1>
        </div>`
      )
      assertCode(codegenElementNode(el))
    })
  })
})

