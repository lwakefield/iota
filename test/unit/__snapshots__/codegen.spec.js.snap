exports[`Codegen codegenChildren generates children 1`] = `"[vnode(\'p\', {}, []), vnode(\'span\', {}, []), vnode(\'h1\', {}, [])]"`;

exports[`Codegen codegenElement generates for an element with attributes and no children 1`] = `
"vnode(\'div\', {
    directives: {
        id: attr(\'id\', \`foo\`),
        class: attr(\'class\', \`class1 class2 \${class3}\`)
    }
}, [])"
`;

exports[`Codegen codegenElement generates for an element with children and attributes 1`] = `
"vnode(\'div\', {
    directives: {
        id: attr(\'id\', \`foo\`),
        class: attr(\'class\', \`class1 class2 \${class3}\`)
    }
}, [vnode(\'p\', {}, []), vnode(\'span\', {}, []), vnode(\'h1\', {}, [])])"
`;

exports[`Codegen codegenElement generates for an element with children and no attributes 1`] = `"vnode(\'div\', {}, [vnode(\'p\', {}, []), vnode(\'span\', {}, []), vnode(\'h1\', {}, [])])"`;

exports[`Codegen codegenElement generates for an element with i-for 1`] = `
"...loop(vals, (val, $index) => vnode(\'div\', {
    key: $index
}, []))"
`;

exports[`Codegen codegenElement generates for an element with i-for and i-if 1`] = `
"...loop(vals, (val, $index) => toggle ? vnode(\'div\', {
    key: $index
}, []) : null)"
`;

exports[`Codegen codegenElement generates for an element with i-if 1`] = `"toggle ? vnode(\'div\', {}, []) : null"`;

exports[`Codegen codegenElement generates for an element with no children and no attributes 1`] = `"vnode(\'div\', {}, [])"`;

exports[`Codegen codegenOptions generates directives 1`] = `
"{
    directives: {
        foo: {
            name: \'foo\',
            value: undefined,
            constructor: Foo
        }
    }
}"
`;

exports[`Codegen codegenOptions generates directives 2`] = `
"{
    directives: {
        foo: {
            name: \'foo\',
            value: bar,
            constructor: Foo
        }
    }
}"
`;

exports[`Codegen codegenOptions generates events 1`] = `
"{
    directives: {
        \'@input\': event(\'input\', $event => {
            const res = handleInput
            return res instanceof Function ? res.call(this, $event) : res
        })
    }
}"
`;

exports[`Codegen codegenOptions generates events 2`] = `
"{
    directives: {
        \'@input\': event(\'input\', $event => {
            const res = handleInput(id)
            return res instanceof Function ? res.call(this, $event) : res
        })
    }
}"
`;

exports[`Codegen codegenOptions generates events 3`] = `
"{
    directives: {
        __formBinding: event(\'input\', $event => name = $event.target.value),
        \'@input\': event(\'input\', $event => {
            const res = handleInput
            return res instanceof Function ? res.call(this, $event) : res
        }),
        value: attr(\'value\', \`\${name}\`)
    }
}"
`;

exports[`Codegen codegenOptions generates multiple attributes 1`] = `
"{
    directives: {
        id: attr(\'id\', \`foo\`),
        class: attr(\'class\', \`class1 class2 \${class3}\`)
    }
}"
`;

exports[`Codegen codegenOptions generates props 1`] = `
"{
    props: {
        foo: foobar
    }
}"
`;

exports[`Codegen codegenTextNode generates tnode 1`] = `"tnode(\`foo \${bar}\`)"`;

exports[`undefined 1`] = `"function sandboxed() {/* istanbul ignore next */++cov_qu0s95ps3.f[3];++cov_qu0s95ps3.s[7];return (/* istanbul ignore next */sandbox.call. /* istanbul ignore next */apply(sandbox, /* istanbul ignore next */[this].concat(values)));}"`;
